File: packages/scriba/src/local/cached.ts
import { resolve } from 'node:path'
import type { ScribaCache } from '../cache/sqlite.ts'
import type { ProviderId } from '../schema/model.ts'
import { addClaudeParsedFileStats, defaultClaudeProjectDirs, parseClaudeFile } from './claude.ts'
import { defaultCodexSessionDirs, parseCodexFile } from './codex.ts'
import { fileFingerprint, isDirectory, walkJsonlFiles } from './files.ts'
import {
	addScannerStats,
	emptyScannerStats,
	type LocalUsageEvent,
	type ScannerStats,
} from './types.ts'

export type CachedCodexEventOptions = {
	cache: ScribaCache
	paths?: string[] | undefined
	env?: Record<string, string | undefined> | undefined
	stats?: ScannerStats | undefined
}

export type CachedClaudeEventOptions = CachedCodexEventOptions

const CLAUDE_PROVIDER_ID: ProviderId = 'claude'
const CODEX_PROVIDER_ID: ProviderId = 'codex'

export async function* iterateCachedClaudeEvents(
	options: CachedClaudeEventOptions,
): AsyncGenerator<LocalUsageEvent> {
	const stats = options.stats ?? emptyScannerStats()
	const seen = new Set<string>()
	const dirs = options.paths?.map((path) => resolve(path)) ?? defaultClaudeProjectDirs(options.env)

	for (const dir of dirs) {
		if (!(await isDirectory(dir))) {
			stats.missingDirectories.push(dir)
			continue
		}

		for await (const filePath of walkJsonlFiles(dir)) {
			const fingerprint = await fileFingerprint(filePath)
			const cached = options.cache.loadFileEvents<LocalUsageEvent>(
				CLAUDE_PROVIDER_ID,
				filePath,
				fingerprint,
			)
			const parsed = cached ?? (await parseClaudeFile(dir, filePath))
			if (cached == null) {
				options.cache.saveFileEvents(
					CLAUDE_PROVIDER_ID,
					filePath,
					fingerprint,
					parsed.events,
					parsed.stats,
				)
			}
			addClaudeParsedFileStats(stats, parsed.stats)
			for (const event of parsed.events) {
				if (event.uniqueKey != null && seen.has(event.uniqueKey)) {
					stats.duplicates += 1
					continue
				}
				if (event.uniqueKey != null) {
					seen.add(event.uniqueKey)
				}
				stats.events += 1
				yield event
			}
		}
	}
}

export async function* iterateCachedCodexEvents(
	options: CachedCodexEventOptions,
): AsyncGenerator<LocalUsageEvent> {
	const stats = options.stats ?? emptyScannerStats()
	const dirs = options.paths?.map((path) => resolve(path)) ?? defaultCodexSessionDirs(options.env)

	for (const dir of dirs) {
		if (!(await isDirectory(dir))) {
			stats.missingDirectories.push(dir)
			continue
		}

		for await (const filePath of walkJsonlFiles(dir)) {
			const fingerprint = await fileFingerprint(filePath)
			const cached = options.cache.loadFileEvents<LocalUsageEvent>(
				CODEX_PROVIDER_ID,
				filePath,
				fingerprint,
			)
			const parsed = cached ?? (await parseCodexFile(dir, filePath))
			if (cached == null) {
				options.cache.saveFileEvents(
					CODEX_PROVIDER_ID,
					filePath,
					fingerprint,
					parsed.events,
					parsed.stats,
				)
			}
			addScannerStats(stats, parsed.stats)
			for (const event of parsed.events) {
				yield event
			}
		}
	}
}

File: internal/cached/cached.go
package cached

import (
	"path/filepath"

	"github.com/agensfield/scriba/internal/cache"
	"github.com/agensfield/scriba/internal/local"
	"github.com/agensfield/scriba/internal/local/claude"
	"github.com/agensfield/scriba/internal/local/codex"
	"github.com/agensfield/scriba/internal/model"
)

func ScanClaude(c *cache.Cache, paths []string) ([]model.LocalUsageEvent, model.ScannerStats, error) {
	stats := model.ScannerStats{}
	var events []model.LocalUsageEvent
	seen := map[string]struct{}{}
	for _, dir := range pathsOr(paths, local.DefaultClaudeProjectDirs()) {
		abs, _ := filepath.Abs(dir)
		if !local.IsDirectory(abs) {
			stats.MissingDirectories = append(stats.MissingDirectories, abs)
			continue
		}
		files, err := local.WalkJSONLFiles(abs)
		if err != nil {
			return events, stats, err
		}
		for _, file := range files {
			fp, err := local.FileFingerprint(file)
			if err != nil {
				return events, stats, err
			}
			parsed, fileStats, ok, err := c.LoadFileEvents("claude", file, fp.Size, fp.MtimeMs)
			if err != nil {
				return events, stats, err
			}
			if !ok {
				parsed, fileStats, err = claude.ParseFile(abs, file)
				if err != nil {
					return events, stats, err
				}
				if err := c.SaveFileEvents("claude", file, fp.Size, fp.MtimeMs, parsed, fileStats); err != nil {
					return events, stats, err
				}
			}
			addStats(&stats, fileStats)
			for _, event := range parsed {
				if event.UniqueKey != "" {
					if _, exists := seen[event.UniqueKey]; exists {
						stats.Duplicates++
						continue
					}
					seen[event.UniqueKey] = struct{}{}
				}
				stats.Events++
				events = append(events, event)
			}
		}
	}
	return events, stats, nil
}

func ScanCodex(c *cache.Cache, paths []string) ([]model.LocalUsageEvent, model.ScannerStats, error) {
	stats := model.ScannerStats{}
	var events []model.LocalUsageEvent
	for _, dir := range pathsOr(paths, local.DefaultCodexSessionDirs()) {
		abs, _ := filepath.Abs(dir)
		if !local.IsDirectory(abs) {
			stats.MissingDirectories = append(stats.MissingDirectories, abs)
			continue
		}
		files, err := local.WalkJSONLFiles(abs)
		if err != nil {
			return events, stats, err
		}
		for _, file := range files {
			fp, err := local.FileFingerprint(file)
			if err != nil {
				return events, stats, err
			}
			parsed, fileStats, ok, err := c.LoadFileEvents("codex", file, fp.Size, fp.MtimeMs)
			if err != nil {
				return events, stats, err
			}
			if !ok {
				parsed, fileStats, err = codex.ParseFile(abs, file)
				if err != nil {
					return events, stats, err
				}
				if err := c.SaveFileEvents("codex", file, fp.Size, fp.MtimeMs, parsed, fileStats); err != nil {
					return events, stats, err
				}
			}
			addStats(&stats, fileStats)
			events = append(events, parsed...)
		}
	}
	return events, stats, nil
}

func pathsOr(paths, fallback []string) []string {
	if len(paths) > 0 {
		return paths
	}
	return fallback
}

func addStats(target *model.ScannerStats, stats model.ScannerStats) {
	target.Files += stats.Files
	target.Bytes += stats.Bytes
	target.Lines += stats.Lines
	target.Events += stats.Events
	target.InvalidLines += stats.InvalidLines
	target.MissingDirectories = append(target.MissingDirectories, stats.MissingDirectories...)
}
